File: 383/383.java
// class Solution {
//     public boolean canConstruct(String ransomNote, String magazine) {
//         for (int i = 0; i < ransomNote.length(); i++) {
//             char r = ransomNote.charAt(i);

//             int matchingIndex = magazine.indexOf(r);

//             if (matchingIndex == -1 ) {
//                 return false;
//             }
//             magazine =magazine.substring(0, matchingIndex) + magazine.substring( matchingIndex +1 );
//         }
//         return true;   
//     }
// }

class Solution {
    public boolean canConstruct(String ransomNote, String magazine) {
        HashMap<Character, Integer> magazineLetters = new HashMap<>();

        for (int i = 0; i < magazine.length(); i++) {
             char m = magazine.charAt(i);

            int currCount = magazineLetters.getOrDefault(m, 0);
            magazineLetters.put(m, currCount + 1);
        }

        for (int i = 0; i < ransomNote.length(); i ++) {
            char r = ransomNote.charAt(i);

            int currCount = magazineLetters.getOrDefault(r, 0);

            if ( currCount == 0) {
                return false;
            }
            magazineLetters.put(r, currCount - 1);
        }
        return true;
    }
}

File: Easy/383/383.js
// /**
//  * @param {string} ransomNote
//  * @param {string} magazine
//  * @return {boolean}
//  */
// var canConstruct = function(ransomNote, magazine) {
//     for ( let i = 0; i < ransomNote.length; i++) {
//         let r = ransomNote[i];
//         let matchingIndex = magazine.indexOf(r);

//         if (matchingIndex === -1) {
//             return false;
//         }

//         magazine = magazine.slice(0, matchingIndex) + magazine.slice(matchingIndex + 1)
//     }
//     return true;
// };

/**
 * @param {string} ransomNote
 * @param {string} magazine
 * @return {boolean}
 */
var canConstruct = function(ransomNote, magazine) {
    let charCount = {};

    for (let char of magazine ){
        charCount[char] = (charCount[char] || 0) +1;
    }

    for (let char of ransomNote) {
        if (!charCount[char] || charCount[char] === 0) {
            return false;
        }
        charCount[char]--;
    }
    return true;
}